### PR TITLE
script: Generate release notes based on GitHub PRs and tags

### PR DESCRIPTION
**Pull request by garrett**
Uploading first draft. There's more work to be done, but it's already working enough for demo purposes.

GOAL:

Have a script that iterates through our repos, finds the release-notes tag, extract the text, and make a draft version of our release notes, ready to edit.

TODO:

- [x] Use the API instead of scraping to get the issues (if possible)
- [x] Split the body text at some predefined break points
   - first use of "release notes", "---", and/or something like that... probably with multiple, with a fallback of just using the body notes
- [x] Get it to "write" a blog post as a file, in a branch, ready for pushing
- [x] Extract the images and rewrite the URLs
- [x] Normalize the heading level (h2 vs h3)
- [x] Handle underline-style titles (for the first title only)
- [x] Document how to use this
   - Comments at the top of the files now suggest running through the toolbox-ruby script, simplifying the process, in #432

Pending PR with generated files: https://github.com/cockpit-project/cockpit-project.github.io/pull/431
Preview: https://garrett.github.io/cockpit-project.github.io/blog/cockpit-248.html

### CODE REVIEW

**Comment by garrett** [comment]
Current output (below the HR):

---

https://github.com/cockpit-project/cockpit/pulls?q=is%3Apr+label%3Arelease-note
## metrics: Install redis package on demand

Always show the "Export to network" switch. Install the redis package if


redis.service does not exist.





Change the pmlogger existence check from `.state === undefined` to


`.exists`, as that is easier to understand, and we were already using


the latter for pmproxy.





Rename TestCockpitPcp to TestMetricsPackages, as this covers more


packages than just cockpit-pcp now.





-----





I interactively tested that with





      systemctl stop pmlogger pmproxy redis; rpm -e --verbose cockpit-pcp pcp redis; systemctl daemon-reload





and then enabling pmlogger and pmproxy either one by one, or both together. After pmlogger and relogin (as advertised), history appears, and after pmproxy, querying metrics with





    curl 'http://localhost:44322/series/labels?names=hostname'





works. Our integration tests can only test package install and actual service configuration separately.





 - [x] Builds on top of #15998


 - [x] Builds on top of #16018





There isn't any thing visually/CSS-y new here, but screenshots for reference.





## Metrics: Install packages on demand





[Cockpit 247](https://cockpit-project.org/blog/cockpit-247.html) introduced setting up a machine for Grafana metrics. However, this only worked if the `pcp` and `redis` packages were already installed. With this Cockpit version, the options will always work, and missing packages get installed on demand.





image





image




## cockpit-session: drop PAM module

Stop the split-out of the client certificate authentication to a PAM


module.  The original reasons for doing so are no longer valid, and the


module design complicates passing additional information into the


authentication step.





The bulk of the logic in the PAM module is moved to a new file,


cockpit-session-client-certificate.[ch].  The PAM module is left around


as a stub, for those who still include it in their configs, but it is


removed from our PAM configs, and the manpage has been dropped.





Some adjustments are made: we use the debug() macro from session-utils


and generally move over to logging to stderr instead of syslog.  We also


replace pam status codes with success bools.





TODO:


 - [x] packaging support for migrating existing configs


 - [x] add back the docs somewhere appropriate





Release note:





## PAM: Deprecate `pam_cockpit_cert` module





The `pam_cockpit_cert` module is no longer required for client certificate authentication, and has been deprecated.  This release continues to ship a stub module which returns `PAM_IGNORE`, but this stub will be dropped in a future release.  If you modified  `/etc/pam.d/cockpit`, the package upgrade will not automatically update the file -- please remove the relevant entry manually in this case.

## bridge: Add channel redirection

This adds a "redirect" option to the open command, which takes the name


of the channel the output of the new one will be redirected to. The


redirect is effectively applied as close to the channels as possible.





Closes #14560





---





There are two cases in which the output is redirected from a channel:


- The output is redirected from a local channel. This is handled in `CockpitRouter` by setting the `redirect_target` property of `CockpitChannel`, which uses it to send data to the redirect target instead of the transport in `cockpit_channel_actual_send`.


- The output is redirected from a peer. This is handled in `cockpit_peer_handle` by passing the redirect target as an argument and then by putting the redirect into a hash table with the (source) channel name as a key. `on_other_recv` then looks up the channel name in the table, again sending it to the redirect target instead of the transport if a corresponding entry is present.





Similarly there are two redirect target types: a local channel and a peer. These are handled by different implementations of the `CockpitRedirect` interface - `CockpitChannelRedirect` for local channels (holding a reference to the corresponding `CockpitChannel` object) and `CockpitPeerRedirect` for peers (holding a reference to the transport and the channel name).





Redirection of a channel is attempted on all bridges - the shortest path then takes precedence (e.g. if there is a redirect between two channels on the same peer, it happens on both the parent bridge and the peer, but effectively the data is redirected on the peer and nothing arrives on the parent). This also means it is compatible with older versions of cockpit-bridge - they simply ignore the redirect option and the redirect is done on the parent bridge (this can obviously be slower, but it will work).





I hope this makes sense - it seems like a small, but useful feature to me. (For potential uses see the linked issue.)







## storage: Integrate encryption better

Cockpit used to expose the device mapper device that implements encryption.  Now we fold that away and pretend the content sits directly in the backing device.





First demo: https://www.youtube.com/watch?v=DASfi39YGZ8


Second demo: https://www.youtube.com/watch?v=kSYuO0WZHWU


Third demo: https://www.youtube.com/watch?v=VnzG2VcDyik





- [ ] #15968 


- [ ] #15937 





Release notes:





### Storage: Encryption is presented as a property of a Filesystem





Cockpit used to make the encryption layer of a encrypted filesystem explicit: It had its own row in the content table and it would be referred to in other places in the UI.  Now the encryption layer is hidden away and Cockpit pretends encryption is a inherent property of a Filesystem.





[ screenshot once the polish is merged ]




https://github.com/cockpit-project/cockpit-machines/pulls?q=is%3Apr+label%3Arelease-note
## Live migration

Transport live VM to another host. Configurable options are:


- Destination URI


- Storage: User can either choose to copy all read-only disks, or they


  can choose that they have a shared storage setup where both source and


  destination hosts have access too. In the latter that case, no disks


  are transported.


- Migrated VM's state: This defines if VM at the destination is running


  or not


- Temporary move: This defines if the migrated VM at the destination


  will be only transient or persistent





Release notes


===========





Cockpit Machines: Migration of a virtual machine


-----------------------------------------------------------------





VM migration is now supported and exposing various configuration options. The user has the ability to choose between live and offline migration, can decide if they want to copy the disks or use a shared storage and lastly there is the option to make the migration temporary.





Screen Shot 2021-05-12 at 17 39 29


Screen Shot 2021-05-12 at 17 42 51







https://github.com/cockpit-project/cockpit-podman/pulls?q=is%3Apr+label%3Arelease-note
## feat: Network options for run images

Add network options in "Run image" dialog, as screenshot:


image





Add two options:





* **Network Namespace Mode**: for selecting `--network` option in podman CLI/`netns.nsmode` in podman API, which allows `bridge`/`none`/`host` and `slirp4netns` (if detected in `podman info`) in rootful mode (system flag), or `none`/`slirp4netns` (if detected in `podman info`) in rootless mode


* **CNI Networks**: for selecting CNI networks using in bridge mode, which will listing CNI networks from `/libpod/networks/json` (`podman network ls` in CLI). Without choosing this option, podman will use default network.





also solved #176





More TODOs for the network feature:





* support more ns modes and options, like `container:id`, `ns:path`, etc.


* support shows IP infos from multiple networks (from `NetworkSettings.Networks`)

## Add import and export of checkpoints

Adds export (download of checkpoint tarball after making a checkpoint) and import (upload of checkpoint tarball for restore) support.





Upload is done by directly reading a file selected by the PatternFly file upload component; ~download is handled using a service worker (it's a little bit hacky, but seems to work fine in both Firefox and Chromium and doesn't store the tarball in memory; maybe there is a better way, but as far as I know Cockpit can download only files from the file system, not channels/HTTP streams)~ download uses channel redirection to save the checkpoint into a file, which is then downloaded directly using a read channel. (Direct download of the checkpoint stream is not possible, because the checkpoint process takes some time, and also errors have to be handled in the UI, not downloaded instead of the tarball).





There are currently no tests - I'm not sure how Selenium handles downloads (the download has to be done in the test browser because of the service worker); I can have a look at it.





checkpoint


restore




https://github.com/cockpit-project/cockpit-ostree/pulls?q=is%3Apr+label%3Arelease-note

**Comment by garrett** [comment]
Goals:

- [ ] Get it to "write" a blog post as a file, in a branch, ready for pushing
- [ ] Split the body text at some predefined break points
   - first use of "release notes", "---", and/or something like that... probably with multiple, with a fallback of just using the body notes
- [ ] Extract the images and rewrite the URLs

**Comment by martinpitt** [comment]
Nice, thanks!  The [VM migration](https://github.com/cockpit-project/cockpit-machines/pull/15) is a special case -- we landed *that*  PR long ago, but didn't announce it yet because there's still outstanding follow-ups to be done. At this point we should probably say "screw it" and announce it anyway in the next release.

With the current way we use the release-note label, the script should by default only look at closed PRs. The open ones have not landed, most of them won't for the next release, and many/most of them don't even have a  proper release note draft on them. The way to say "this PR will go into the next release" is an additional [release-blocker label](https://github.com/cockpit-project/cockpit/labels/release-blocker). So looking at that in addition would actually give us enough control over what "goes in".

**Comment by garrett** [comment]
> With the current way we use the release-note label, the script should by default only look at closed PRs. The open ones have not landed, most of them won't for the next release, and many/most of them don't even have a proper release note draft on them.

True, but sometimes the PRs don't land until after we start working on the release notes... and this was one of the big points I was making last week.

**Comment by garrett** [comment]
I need to squash this down a bit, but it's probably ready for merging soon. :wink:

**Review by martinpitt — APPROVED** [review]
Thanks for writing this automation! The code looks very reasonable -- I didn't go through it with a too fine comb, as it obviously produces nice results in #431 and we can (and very likely will) always adjust it to our needs.

File: _scripts/generate-release-notes.rb
Comment (garrett) [reply _scripts/generate-release-notes.rb:181]: It's needed for the forks that are not necessarily at a root domain. I think I wrote something in our template to handle it transparently for blog posts though when it's missing.

File: _scripts/generate-release-notes.rb
Comment (garrett) [reply _scripts/generate-release-notes.rb:139]: I want something with a decent default, so it's attributed to whoever runs the script. But it's easy to edit if it's not correct.

File: _scripts/generate-release-notes.rb
Comment (garrett) [reply _scripts/generate-release-notes.rb:177]: We can edit the alt text in the release notes. GitHub doesn't generally give us good alt text by default, so I wanted something that is not broken by default. It handles stylical screenshots and images. I guess I should've also made it case insensitive so "screen shot" also works, but our screen shots with Firefox and GNOME are all "Screen Shot", IIRC, so :shrug: 

If there's _actual_ alt text in the PR that isn't just image or "Screen Shot...", it picks that up and uses it instead. This is just a fallback for a very common set of cases where the alt text isn't really specified in the PR.

**Pull request merged**
